Adding class for V0 selection

sartozza opened the pull request.

tklemenz committed:
TPC CDBInterface: pick specific file from CCDB with timestamp and meta data filter + newZSCalib helper function for TPC QC

aphecetche committed:
[O2-2132] CMake: Get rid of custom test wrapper script

In favor of using native cmake/ctest facilities as much as possible.

It's important to note that a couple of things should be changed on the
calling side (o2.sh in the alibuild case) as well _if_ we want to have a
1-to-1 matching with the current behavior :

- if more that one attempt per test use `ctest --repeat until-pass:3`
- no default timeout is set. If that's not adequate use `ctest --timeout 100`

The (cmake) rationale for the timeout is most probably dependent on the
machine the test is ran on, so it's hard to put a default timeout value
in any CMakeLists.txt

robincaron13 committed:
Add the geometry misaligner class

badarots committed:
Initial attempt on TPC dEdx calibration

robincaron13 committed:
Add small fixes and comments

rpezzi committed:
Fix CI warnings

robincaron13 committed:
Add CCDB path to store AlignParams objects

robincaron13 committed:
Add external input misalign parameters

Kavaldrin committed:
fix unmatched input/output

robincaron13 committed:
Fix alignID and not applying alignment to geom

robincaron13 committed:
Add test macro and small fixes in misaligner

alibuild committed:
Please consider the following formatting changes

aferrero2707 committed:
[MCH] added decoder benchmarking

aferrero2707 committed:
[MCH] improved code structure

This commit does not introduce any new feature,
it only aims at improving the code readability

aferrero2707 committed:
[MCH] digits merging implementation

This commit introduces the code that merges digits
that belong to the same detector signal.

The merging is performed when the first ADC sample
of a given digit follows immediately the last ADC sample
of an existing digit for the same readout channel.

preghenella committed:
Add Pythia8 process name and code to MC event header

aferrero2707 committed:
[MCH] make sure the digits merge table is allocated on the heap

This also fixes a segfault in the ClosureCoDec test

matthiasrichter committed:
Make raw page data generator for DPLUtils unit tests a separate class

There will be more unit tests and micro benchmarks using the data
generator.

martenole committed:
TRD calibration for vDrift and ExB added

- results are written to CCDB

fcolamar committed:
PWGHF: Improved MC-reco mode of DDbar correlation code

fcolamar committed:
Fix to comment lines

fcolamar committed:
Added support for reflection mass distributions

mpuccio committed:
Move to proper logger + protect file reads

HauenderPhilip committed:
Changed the Kr-ClusterFinder such that it can read arbitrary long digits.

Implemented suggested changes.

badarots committed:
Move TrackCuts to DataFormatsTPC

martenole committed:
Remove unnecessary track seed refs

fcolamar committed:
Added reflection cases also to 2D mass plots and correlations distributions

fcolamar committed:
Adapted cases for D+D- to preserve compatibility with taskCorrelationsDDbar

martenole committed:
Move transient data members from TRD tracks into helper struct

TODO: take care of GPU track types and continuous tracking mode

mpuccio committed:
Fix json parsing and clang-tidy error

Co-authored-by: Stefano Trogolo <strogolo@users.noreply.github.com>

martenole committed:
Add path length support, not sure where calculation should be done

ktf committed:
DPL: export environment variables when dumping DDS config.

Commit:
Cables to the MFT and FIT patch panel

bazinski committed:
trd fix corrupt data at end of digits

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamTrackSelection.h:
```diff
@@ -281,6 +288,9 @@ bool FemtoDreamTrackSelection::isSelectedMinimal(T const& track)
   if (nDCAzMaxSel > 0 && std::abs(dcaZ) > dcaZMax) {
     return false;
   }
+  if (nDCAMinSel > 0 && std::abs(dcaXY) < dcaMin && std::abs(dcaZ) < dcaMin) {
```
shouldn't we add them quadratically?

sartozza replied:
Yes you are right actually.
Corrected by defining:
  const auto dcaXYZ = std::sqrt(pow(dcaXY, 2.) + pow(dcaZ, 2.));
nDCAMinSel > 0 && std::abs(dcaXYZ) < dcaMin

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamTrackSelection.h:
```diff
@@ -370,6 +384,7 @@ void FemtoDreamTrackSelection::fillQA(T const& track)
     mHistogramRegistry->fill(HIST("TrackCuts/tpcnsharedhist"), track.tpcNClsShared());
     mHistogramRegistry->fill(HIST("TrackCuts/dcaXYhist"), track.pt(), track.dcaXY());
     mHistogramRegistry->fill(HIST("TrackCuts/dcaZhist"), track.pt(), track.dcaZ());
+    mHistogramRegistry->fill(HIST("TrackCuts/dcahist"), track.dcaXY());
```
why dcaXY here?

sartozza replied:
Corrected with sqrt(dcaXY^2 + dcaZ^2)

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamTrackSelection.h:
```diff
@@ -477,6 +493,11 @@ void FemtoDreamTrackSelection::fillCutQA(T const& track, uint64_t cutContainer)
             break;
+          case (femtoDreamTrackSelection::kDCAMin):
+            if (isTrue) {
+              mHistogramRegistry->fill(HIST("TrackCutsQA/dcaMin"), currentTrackSelCounter, dcaXY);
```
same here, should be sqrt(dcaXY^2 + dcaZ^2)

sartozza replied:
Corrected.

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,351 @@
+             kInvMassRejMax };
+enum ChildTracks { kPosDaughTrack,
+                   kNegDaughTrack };
+enum class V0Type { kK0s,
```
why enum class here?

a-mathis replied:
also, we were looking into this o2::track::PID, where this is already defined

sartozza replied:
By reading the updates applied in C++11, I found out that enum class is to be preferred over enum since it allows more flexibility in terms of the name of the enum and on the type. 
But I can let it be an enum.

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,351 @@
+  /// basic function to set from the task which kind of V0s has to be analysed
+  void setTypev0(femtoDreamV0Selection::V0Type v0type)
+  {
+    mSelV0 = v0type;
```
I am not sure I understand the logic here - do I need a separate set of cuts for AntiLambdas?

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,351 @@
+             kDecVtxMax,
+             kInvMassV0Max,
+             kInvMassRejMax };
+enum ChildTracks { kPosDaughTrack,
```
ChildTrackType? ChildTrackCharge? Just to make sure that it is clear what you're after here
(also, you have then Daugh in the enum itself xD)

sartozza replied:
ChildTrackType and I remove Daugh.

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,351 @@
+  float mMassV0 = 0.;
+  float mMassRej = 0.;
+  float mMassPDGV0 = 0.;
+  float mMassPDGRej = 0.;
```
why do you need these as members?

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,351 @@
+{
+  if (registry) {
+    mHistogramRegistry = registry;
+    fillSelectionHistogram("V0Cuts/cuthist"); ///For now empty since I need to understand the next nSelections
```
I don't understand the comment here

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,351 @@
+    mHistogramRegistry->add("V0Cuts/invMassPt", "; #it{p}_{T} (GeV/#it{c}); #it{M}_{inv} (GeV/#it{c}^{2})", kTH2F, {{8, 0.3, 4.3}, {40, 1., 1.2}});
+    mHistogramRegistry->add("V0Cuts/cpapTBins", "; #it{p}_{T} (GeV/#it{c}); #it{cos(#alpha)}", kTH2F, {{8, 0.3, 4.3}, {1000, 0.9, 1.}});
+
+    const int nPtV0MinSel = getNSelections(femtoDreamV0Selection::kpTV0Min);
+    const int nPtV0MaxSel = getNSelections(femtoDreamV0Selection::kpTV0Max);
+    const int nDCAV0DaughMax = getNSelections(femtoDreamV0Selection::kDCAV0DaughMax);
+    const int nCPAV0Min = getNSelections(femtoDreamV0Selection::kCPAV0Min);
+    const int nTranRadV0Min = getNSelections(femtoDreamV0Selection::kTranRadV0Min);
+    const int nTranRadV0Max = getNSelections(femtoDreamV0Selection::kTranRadV0Max);
+    const int nDecVtxMax = getNSelections(femtoDreamV0Selection::kDecVtxMax);
+    const int nInvMassV0Max = getNSelections(femtoDreamV0Selection::kInvMassV0Max);
+    const int nInvMassRejMax = getNSelections(femtoDreamV0Selection::kInvMassRejMax);
```
not needed? -> remove

sartozza replied:
Yes this was a residuals of the FillcutQA which I removed.

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,351 @@
+}
+
+template <typename C, typename V>
+bool FemtoDreamV0Selection::isSelectedMinimal(C const& col, V const& v0)
```
this function should also call FemtoDreamTrackSelection::isSelectedMinimal for the child tracks

sartozza replied:
Check for tracks added:
  if (!posTrack.isSelectedMinimal(posTrack)) {
    return false;
  }
  if (!negTrack.isSelectedMinimal(negTrack)) {
    return false;
  }

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,351 @@
+  ClassDefNV(FemtoDreamV0Selection, 1);
+}; // namespace femtoDream
+
+void FemtoDreamV0Selection::init(HistogramRegistry* registry)
```
you also need to init the child tracks here

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,351 @@
+      mMassPDGV0 = RecoDecay::getMassPDG(3122);
+      mMassRej = v0.mK0Short();
+      mMassPDGRej = RecoDecay::getMassPDG(310);
+      break;
```
but remember, we were looking into this o2::PID thingy, which sets the mass properly

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,351 @@
+  return true;
+}
+
+template <typename T1, typename T2>
```
I think this function is short enough to go to the declaration

sartozza replied:
Moved to declaration.

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,351 @@
+  uint64_t output = 0;
+  size_t counter = 0;
+  const auto signPos = posTrack.sign();
+  const auto signNeg = negTrack.sign();
+  if (signPos < 0 || signNeg > 0) {
+    printf("ERROR - Wrong sign for V0 daughters");
+  }
```
maybe check that already for the minimal selection?

sartozza replied:
Moved to the minimal selection function.

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,351 @@
+  const auto cpav0 = v0.v0cosPA(col.posX(), col.posY(), col.posZ());
+  const std::vector<float> decVtx = {v0.x(), v0.y(), v0.z()};
+
+  switch (mSelV0) {
```
also here the task should be capable of doing multiple particles at the same time

a-mathis reviewed (commented):
Thanks a lot for this development! I have some minor comments, and one major - for now the task only allows one species at a time. For to my opinion, the task should be capable of running Lambda/Antilambda at the same time (+ even K0 eventually)

martenole committed:
Add filter for trigger records without ITS contribution

bazinski committed:
move tracklet HCH to beginning of tracklets and always present

shahor02 committed:
Make decoding errors treatment more robust

afurs committed:
FITRaw: hotfix for #6562 (#6566)

njacazio committed:
Add first task of the LUT maker from fullsim (#6498)

- Add single particle QA task
- Add 2D efficiency to efficiency QA task
- Add possibility to select charge depending on PDG code
- Add Y for MC particles

ktf committed:
DPL: dump metrics at interval if enabled (#6526)

matthiasrichter committed:
Do not consider type {DPL/EOS} as unknown in the raw proxy

DPL output proxy automatically add message of type {DPL/EOS} to forward
end-of-stream on all output channels. If another DPL workflow connects
via raw proxy, it handles the channel info headers and can ignore the
actual message.

Also adding an option to test_ExternalFairMQDeviceWorkflow which allows
to drop the subscriber part of the test and instead connect the raw proxy.

sawenzel committed:
Fix division by zero in MFT TrackFitter

This is avoiding potential division-by-zero exceptions (seen
on MacOS) by replacing the division with a signum operation.

shahor02 committed:
Fix obsolete copyright headers

sawenzel committed:
AOD conversion: Correct indexing of MCparticle to MCCollision

MCparticles referred to the original eventIDs
from the Geant transport, whereas they should refer
to the MCcollision AOD table entry.

Thanks to Nazar for discussions.

shahor02 committed:
fix in RDH update for CRU det. in trig. mode

ktf committed:
DPL: expire resource offers (#6549)

aalkin committed:
[O2-2424] DPL Analysis: Rework slicing algorithm for merged AODs (#6557)

* Rework slicing algorithm for merged AODs solving O2-2424

davidrohr committed:
FST: Add multiplicities for TRD components to be fast enough

davidrohr committed:
FST: Increase sleeps during memory allocation to avoid wrong NUMA pinning

aphecetche committed:
[MRRTF-131] MCH: Add reader and writer specs for TrackMCH

Also add one feature to o2-mch-tracks-file-dumper : the ability to read
from Root files as well.

aferrero2707 committed:
[MCH] added check for DS channel overflow

jolopezl committed:
remove TRDBase/Digit.h

ktf committed:
DPL: protect recursive preloading of libSegFault.so

martenole committed:
Update copyright notice

ktf committed:
DPL Analysis: demote message to info

Commit:
Fixing typo on MB support shelf name

shahor02 committed:
Allow passing detector mask in dummy alignment upload macro

sawenzel committed:
Enable GRP semaphore by default / automatic removal

chiarazampolli committed:
Update names of configurables and some comments

clang-format

aalkin committed:
fix filtered grouping for merged AODs

aphecetche committed:
[MRRTF-122] Add MCH to FST

Note that the last tracking part (extrap to vertex) is far from final
and will be updated in a forthcoming PR (in prep) where FwdTracks
will be added to AODs and extrapolation will be part of the AOD
producer.
Currently the vertex information that is used is completely fake (coming
from a vertex sampler that is used for debugging when using Run2 data).

shahor02 committed:
Fixes for TPC-ITS matching in ITS triggered mode

AllaMaevskaya committed:
FT0 module's positions according drawings

AllaMaevskaya committed:
FV0 alignment

sawenzel committed:
Remove remaining pragma once

bazinski committed:
TRD fix for empty frames (#6511)

* trd reading empty frames

* safter sending of blank frames

* ensure empty frame is indeed empty

* flexible data description

* move TRD eventrecord class to reconstruction dir

* remove classdef for EventRecord, and local variables for emptry frames data

sawenzel committed:
A simple github action to check pragma once rule (#6592)

* Create pragmaonce.yml

* Update pragmaonce.yml

martenole committed:
Separate tracklet transformer and tracking and use InputHelper in both workflows

wiechula committed:
TPC: Improve detection/removal of duplicate digits

shahor02 committed:
Store det.mask in gGeoMnager and use in misalignment

gGeoManager->GetUniqueID() will provide the mask of added detectors, only volumes actually present
in the geometry will be misaligned when loading the geometry

aimeric-landou committed:
lambdakzeroanalysis: add configurable for Dca histograms binning (#6556)

victor-gonzalez committed:
First step towards the simulation challenge (#6530)

* First step towards the simulation challenge

For the time being is being a separate task
The final goal is a single one

Implemented
- data collection directly in differential form
- huge decrease in memory layout

* Introduction of the data collecting engine structure

The whole data collecting functionality is encapsulated
into an structure which can be replicated as needed

* Just a single task for the different cent/mult ranges

The idea is to have a multitask scenario for running
systematic tests in 'parallel' exclusively

sstiefel19 committed:
add mc version of gammaTask and event cut histogram (#6588)

ktf committed:
DPL Analysis: correctly track consumed offers (#6600)

sawenzel committed:
Update pragmaonce.yml

sawenzel committed:
Update pragmaonce.yml

sawenzel committed:
o2-sim: Ensure correct order of events in output file

So far, the output event sequence did not
necessarily match the order in whey they
were generated. This is now ensured to avoid
bookkeeping errors.

This comes at slight memory cost, as events
can't be flushed as soon as they arrive.

Minor other changes:
- fix pythia8 call in example

jbarrella committed:
TRD digit converter improvements

aalkin committed:
DPL Analysis: Slice and Array index columns (#6543)

* Introduction of Slice and Array index columns

* (fixed-size) Array index column: getter returns std::array with
  iterators to the bound table
* Slice index column: getter returns a slice of the bound table, defined
  by the values of 0th and 1st elements of the index
* Added a test in test_ASoA.cxx

* transfer index bindings to a new slice

rpezzi committed:
Fix copyright headers

knopers8 committed:
[QC-134] Moving window in Mergers (#6456)

ktf committed:
DPL Analysis: demote shm accounting messages (#6606)

davidrohr committed:
GPU: Add tpcIncreasedMinClustersPerRow option

davidrohr committed:
GPU: Bugfix: buffer size corrections must be performed before clusterization, to affect also buffers of clusterizer

martenole committed:
Check filter flag in TRD reco and adapt FST script

mconcas committed:
Fix some of the ALICE 3 geometry issues (#6562)

* Fix length and thickness of TRK

* Update ALICE 3 beam pipe default length

hahassan7 committed:
[EMCAL-694] Summing all digits within one tower using SDigitizer

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/femtoDreamProducerTask.cxx:
```diff
@@ -149,8 +202,57 @@ struct femtoDreamProducerTask {
+      auto cutContainerV0 = v0Cuts.getCutContainer(col, v0, postrack, negtrack);
+      if ((cutContainerV0.at(0) > 0) && (cutContainerV0.at(1) > 0) && (cutContainerV0.at(2) > 0)) {
+        int postrackID = v0.posTrackId();
+        int rowInPrimaryTrackTablePos = -1;
```
I think it would be nice to make this a tiny function in the task since it is also used for the neg. child

sartozza replied:
The tiny function is referred specifically to which part?
The definition of
int postrackID = v0.posTrackId();
int rowInPrimaryTrackTablePos = -1;
and then the loop over the tracks ID?
 
So it would be a function taking as input the v0, and the vector and returning a int given by the row in the primary tracks table. Something like this?

a-mathis replied:
exactly, something which takes all the necessary ingredients and return the proper ID in the femtoDreamParticleTable referring to the primary track

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/femtoDreamProducerTask.cxx:
```diff
@@ -149,8 +202,57 @@ struct femtoDreamProducerTask {
+            break;
+          }
+        }
+        childIDs[0] = rowInPrimaryTrackTablePos;
```
This is a bit misleading , can you just replace the entry of the pos track?

sartozza replied:
You mean here the negative track right childIDs[1]?
So in the negative daughter track inputs of the table we only have the corresponding value.

a-mathis replied:
yes

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamTrackSelection.h:
```diff
@@ -231,7 +237,7 @@ bool FemtoDreamTrackSelection::isSelectedMinimal(T const& track)
   const auto dcaXY = track.dcaXY();
   const auto dcaZ = track.dcaZ();
-
+  const auto dcaXYZ = std::sqrt(pow(dcaXY, 2.) + pow(dcaZ, 2.));
```
Maybe dca?

sartozza replied:
Done.

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamTrackSelection.h:
```diff
@@ -298,6 +309,8 @@ uint64_t FemtoDreamTrackSelection::getCutContainer(T const& track)
   const auto tpcNClsS = track.tpcNClsShared();
   const auto dcaXY = track.dcaXY();
   const auto dcaZ = track.dcaZ();
+  const auto dcaXYZ = std::sqrt(pow(dcaXY, 2.) + pow(dcaZ, 2.));
```
dca?

sartozza replied:
Done.

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,248 @@
+  /// Initializes histograms for the task
+  void init(HistogramRegistry* registry);
+
+  template <typename T>
```
This is not used anymore, so please remove

sartozza replied:
removed the SetPIDSpecies

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,248 @@
+ private:
+  FemtoDreamTrackSelection PosDaughTrack;
+  FemtoDreamTrackSelection NegDaughTrack;
+  std::vector<o2::track::PID> mPIDspecies;
```
Temove

sartozza replied:
Here I remove only the mPIDspecies since it is not used at all.
The FemtoDreamTrackSelection are used in the class.

a-mathis replied:
perfect, thanks!

a-mathis commented on Analysis/Tasks/PWGCF/FemtoDream/include/FemtoDream/FemtoDreamV0Selection.h:
```diff
@@ -0,0 +1,248 @@
+    fillSelectionHistogram("V0Cuts/cuthist"); ///For now empty since I need to understand the next nSelections
+
+    /// \todo this should be an automatic check in the parent class, and the return type should be templated
+    int nSelections = 2 + getNSelections(); /// the 2 still holds for the first two bits???
```
No 2+ anymore

sartozza replied:
Corrected.

a-mathis reviewed (commented):
Fantastic, this looks great! I have a few minor comments, and please remove additionally the debug output

aalkin committed:
[O2-2453] DPL Analysis: Fix integer result type inference for expressions

martenole committed:
Fix unused variable warning

aalkin committed:
[O2-2452] DPL Analysis: Fix extende tables example to account for unassigned tracks

davidrohr committed:
GPU: Fix integer overflow in buffer size estimation

sawenzel committed:
It's enough to use MCHeader file to create digi context

bazinski committed:
fix sending whole messages

ktf committed:
DPL: process region callbacks as early as possible (#6553)

In case we are not running, we can afford processing the callbacks
as soon as possible.

sawenzel committed:
Less verbose printing in HitMerger

vkucera committed:
CI: Run space checker on pull_request_target

matthiasrichter committed:
Adding DPL parser tool for raw page sequences provided by DPL input.

The DPLRawPageSequencer can be used to find sequences of consecutive
raw pages with a similar property, e.g. the FEE ID. The actual check
and callback to handle a sequence can be provided by lambda functions.

The raw pages within one buffer/message are expected to have the full
length except the last page which can be smaller. The fixed spacing
allows fast scanning by binary search.

Corresponding unit test is emulating variable-length sequences of
raw pages.

fixup! Adding DPL parser tool for raw page sequences provided by DPL input.

matthiasrichter committed:
Minor bugfix, declaring class member storing template feature as static const

aphecetche committed:
[O2-2461] MCHRaw: bc offset should be applied before the rollover

shahor02 committed:
Dont disable FillEmptyHBF for CRU det. in triggered mode

fcolamar committed:
Fixed histogram labels

aalkin committed:
DPL Analysis: make all indices behave consistently with Filtered binding (#6607)

aalkin committed:
DPL Analysis: add unsigned int types to Variant/Configurable (#6577)

* add unsigned int types to Variant

* add handling and tests for uints

nburmaso committed:
Add more information to TrackExtra table (#6596)

* Add TPC information (tpcNClsFindableMinusFound)

* add TOFsignal in TOF matched info

* Fix tofExpMom calculation

* Get expSig via getTOF()

MichaelLettrich committed:
[CTF] Bugfix swaped entries in literals metadata

The Metadata of the encoded CTF data in EncodedBlocks was set wrongly
during the encoding process, causing erronious decoding of
incompressible symbols.
This fix swaps the values for `nLiterals` and `nLiteralWords` to ensure
decoding is correct again.

sawenzel committed:
Option to scale global material density for systematic studies

Use `--configKeyValues SimMaterialParams.globalDensityFactor=xx`

renaming SimCutParams.h -> SimParams.h; This header now
hosts more than one parameter for transport simulation.

arvindkhuntia committed:
[FV0] (1) Add separate waveforms for ring 1-4 and ring 5 and  (2) take care of adc overflow (#6611)

* [FV0] (1) Add separate waveform for ring 1-4 and ring 5 and take care of adc overflow

aphecetche committed:
MCH: Include CH6L in electronic mapping (#6605)

* MCH: Include CH6L in electronic mapping

* remove tabs

ktf committed:
DPL: force processing of region callback without waiting for data (#6621)

Without this, it is possible that a late arriving event will not be processed
until some data actually arrives on one of the channels, unblocking the event
loop. This should obviate to the problem by triggering an aync libuv event
which will unblock the event loop without the need for data.

sawenzel committed:
Catch arrow Check failed/assert to prevent DPL workflow hangs

jgrosseo committed:
Fix order to align with hyperloop (#6633)

cortesep committed:
ZDC Adding protection for missing STF (#6529)

* Fix Digits2Raw missing addData

* Fix compilation error

* Fix problem with linkid

* Fix problem with linkid

* Fix problem with linkid

* Handling missing STF

* Small printout improvements

* Make data members private in RawReaderZDC.h

* Make data members private in RawReaderZDC.h

* Selecting only ZDC raw data stream

badarots committed:
Missing CMake entry in #6305

sevdokim committed:
Fix for CPV/RawReaderMemory to handle HBF orbits properly at the beginning of reading

shahor02 committed:
Do not LOG(INFO) from device configuration

as it pollutes the xml topology preparation

mfasDa committed:
[EMCAL-630] Move to gamma2 raw fitter by default

noferini committed:
add DistStf as mandatory input for TOF decoder (#6625)

* add DistStf as mandatory input for TOF decoder

* filtering distStf input in TOF decoder loop

* add dist stf in tof compressor as input

mconcas committed:
[GPU/Full System Test] GPU standalone benchmarking (#6484)

* Add CUDA backbone

* HIP breaks

* Make two separate libraries

* Re-arrange directories

* Add missing header

* Meta library does not compile

* Port hipInfo example to test gpu specs

* Fix compilation to test build on EPN

* Produce two separate executables

* Flatten dir tree a bit

* Cleanup

* Add CMake forced re-configuration

* HIP can't find symbols

* Checkpoint before radical change

* Create single executable

* Update

* Add first dummy benchmark

* Assign a block to each scratch segment

* Fix copyright

* Please consider the following formatting changes (#16)

* Set configurable iterations

* Improve busy fucntion + streaming results on file

* Fix bug in CLI params

* Add configurable number of tests

* Fix undefined behaviour insetting nLaunches

* Please consider the following formatting changes (#17)

* Streamline ker benchmarking w/ events

* Tidy up kernels and improve output

* Update read test

* Please consider the following formatting changes (#18)

* CP

* Add last read test

* Add last read test

* Fix result dump on file

* add reading kernel

* Add write tests

* Add copy benchmark

* Remove CommonUtils dependency

* Remove GPUCommon dependency

* Fix fullCI errors

* Revise result saving

* Ready to test on EPN

Co-authored-by: ALICE Builder <alibuild@users.noreply.github.com>

AllaMaevskaya committed:
use Nchannels as FT0 constants

nburmaso committed:
Add V0s table to the converter

bazinski committed:
modify endpoint definition and sort tracklets before sending

sawenzel committed:
A utility to engineer collision structures in a timeframe

Going beyond what was possible until now
via SimReader:

* make collision contexts of arbitrary timelength

  `o2-steer-colcontexttool -i bkg,50000 --tflength 10`
  (generated a context for 10ms timeframes and 50kHz interaction rate)

* allow to inject only into every N-th background event

  `o2-steer-colcontexttool -i bkg,50000 sgn,@0:e5`
  (injects signals into every 5-th background event)

* allow to inject respecting a minimal time distance between signal events

  `o2-steer-colcontexttool -i bkg,50000 sgn,@0:d10000`
  (injects signals into background but minimally 10000ns apart)

* allow to overlay arbitrary interactions (without matching of vertex)

  `o2-steer-colcontexttool -i bkg,50000 sgn,40000`
  (2 unrelated collisions, the first one at 50kHz the second one at 40kHz)

* allow to inject multiple signal events

  `o2-steer-colcontexttool -i bkg,50000 sgn1,@0:e1 sgn2,@0:e1`
  (injects 2 signals into every background event)

The tool can be upgraded to allow editing of existing contexts, etc. in the future.
One can also treat QED events in this new fashion without special treatment in digitizers.

To use the tool, simply generate a context with it before digitization
```
o2-sim -n ... -o bkg
o2-sim ... -o sgn
o2-steer-colcontexttool -i bkg,40000 sgn,@0:e2
o2-sim-digitizer-workflow --incontext collisioncontext.root --sims bkg
```

nburmaso committed:
Use GlobalTrackID as a key for unordered_map instead of int index

nburmaso committed:
Add cascades to the converter

nburmaso committed:
Add an exception throw if positive or negative track index not found

ktf committed:
DPL: fix idling readers

No need to empty the queue at End Of Stream when devices do not
have any data inputs. We can simply switch to the "Idle" state and
hopefully not use any CPU.

nburmaso committed:
Add an exception throw if a bachelor track index not found

TimoWilken committed:
Export Framework headers required for O2Physics (#6641)

In order to factor out O2Physics into its own repository, some extra O2 headers
need to be exposed.

So far, these are only the headers required for
`Analysis/Tutorials/src/histograms.cxx` -- copying other analysis code might
require more headers to be exposed.

mpuccio committed:
Make json workflow description avalable for cefp (#6623)

anerokhi committed:
DPL GUI: fix switch-case fallthrough (#6646)

ktf committed:
DPL Analysis: better tweaking of default aod-memory-rate-limit (#6645)

shahor02 committed:
Extend detectors CTFHeader by format version and dict. timestamp

anerokhi committed:
MathUtils: Fix a copy-paste bug introduced in ae8838eec8c3148fef407725f69b416251c9ff8f

anerokhi committed:
DPL: fix unassigned variable (#6647)

cortesep committed:
Bug fix from @anerokhi

sawenzel committed:
Fix typo in RecoContainer concerning V0s

The input spec should be all capitalized `V0S` to be consistent with output spec definitions.

shahor02 committed:
full_system_test.sh can run with externally provided CTF dictionary

if called as
CREATECTFDICT=0 $O2_ROOT/prodtests/full_system_test.sh

fgrosa committed:
PWGHF: Rearrange selections and add possibility to apply pT-differential cuts (#6392)

anerokhi committed:
[MID] remove unwanted local scope of variable (#6649)

mconcas committed:
Set configurable GPU device from CLI

Set device number in outfile name

shahor02 committed:
thread-safe version of Gaussian fit with error extraction

anerokhi committed:
TPCReconstruction: Fix typo

aalkin committed:
DPL Analysis: add an exception on incorrect join (#6644)

dsekihat committed:
Analysis/PWGDQ: merge v0selector in DQ (#6634)

* Analysis/PWGDQ: merge v0selector in DQ

* Analysis/PWGDQ: remove unnecessary lines

anerokhi committed:
Remove leftover count increment

Looks like it was forgotten when the code switched from range-based for loops b1d7bff365b78cbde6b087376b7704f3d9d6e744
std::string records; seems to have never been used

anerokhi committed:
MatchTPCITS: fix leftover declaration of variable

See dc091dba2e910273d38d0a4dac48963b427f2adf which tried to move the declaration of itsROBin out of the loop

mconcas committed:
Add toggle for test types

shahor02 committed:
fix abs to std::abs in CaloRawFitterGS.cxx

shahor02 committed:
use reference instead of copy in the loop, CTP/Configuration

anerokhi committed:
DPL GUI: Fix format specifiers for (u)int64_t (#6662)

mconcas committed:
Add macro to visualize result

anerokhi committed:
Fix invalid iterator use UB

mCache.pop_front() invalidates iterators to the first element. Here, the current loop iterator value is always iterator to the first element, so, in loop exit condition comparison against past-the-end iterator, the loop iterator will be invalid. Using invalid iterators in comparison may have undefined behavior.

anerokhi committed:
Do not ignore O2_BUILTIN_(UN)LIKELY parameter (#6671)

anerokhi committed:
Add va_end (#6667)

anerokhi committed:
Jet code: Prevent expensive copying (#6665)

fapfap69 committed:
Remove obsolete functions form Geo.cxx

anerokhi committed:
Remove dangling throw (#6669)

anerokhi committed:
Silence a potentially unused variable warning (#6651)

Not sure if the warning is legit, but it's clear enough what one would need to look into.

martenole committed:
TRD raw reader can dump to file

shahor02 committed:
Safer way to erase TimeSlots

njacazio committed:
PWGPP: Update efficiency task (#6656)

- Add efficiency of nuclei
- Add plots of selection
- Fix selection of PDG

wiechula committed:
TPC: make canvas own histograms

In case of calling a draw function which does not direclty return the
hitogram pointers, but it is drawn directly on a canvase, make the
canvas own the histogram, such that it will be deleted when deleting the
canvase or calling Clear().

knopers8 committed:
[QC-110] Performance improvements in Dispatcher (#6560)

The Dispatcher::run() performs 45% faster with these changes for 1000 parts and 10% random sampling.
highlights:
 - match inputs, not parts
 - const ref instead of move ref in Dispatcher::send()
 - reuse the found output route when matching
 - compute deviceID once
 - decide once per input (not part) and policy
 - create DataSamplingHeader once per policy

knopers8 committed:
Return optional port in DataSampling::PortForPolicy (#6282)

* Return optional port in DataSampling::PortForPolicy

* Fix test failure

martenole committed:
Add README for TRD reco workflow

ktf committed:
Fix compilation issue on macOS (#6683)

Requires explicit cast, apparently.

sartozza committed:
"Update femtodream: V0 selection class and Math class" (#6628)

* [o2femtodream] Reader and pair task and corresponding histogramming

o2FemtoDream: adding Math class and update of Container to produce only histograms for Pair

Configuring the init in container for histograms with ConfigurableAxis

Update on FemtoMath

* [o2femtodream] Extend the data type to incorporate more information

* Adding class for V0 selection

* adding V0 selection

Commit:
New Beam Pipe Support on the C side

ktf committed:
DPL: misc cleanups (#6682)

* Fix pessimizing moves
* Cleanup namespace in MessageContext.h

aalkin committed:
DPL Analysis: fix and update table builder (#6676)

* remove include; move non-templated code to cxx
* make sure empty tables keep labels

njacazio committed:
Add description of PID tables (#6678)

anerokhi committed:
Use array delete

mconcas committed:
Test not writing data

sawenzel committed:
adjust jobutils error detection pattern

wiechula committed:
TPC: Update laser track drift velocity calibration

* add device to pre-filter laser track candiates
  (tpc-laser-track-filter)
* split device for
  - simple local processing (tpc-calib-laser-tracks)
  - distributed processing (tpc-laser-tracks-calibrator)
* new struct for calibration data storage
* drift velocity calibration separately for A-/C-Side
* code cleanup

bovulpes committed:
MFT: add tracker option for clusters full scan

mconcas committed:
Destroy events after usage

bazinski committed:
fix linkid and hcid mapping and sorting

mconcas committed:
Destroy stream after usage, thanks

wiechula committed:
TPC: several unrelated smaller fixes and improvements

wiechula committed:
TPC: fix treatment of sync offset in LinkZS data

Ideally, the sync offset is aligned to the largest offset within one TF.
This value can be found by looping over all data once but requires
access to the full TF.
Alternatively, an assumption about the largest sync offset can be made.

Some unrealted fixes are includes as well

cortesep committed:
Fix fp conversion and bc addressing (#6672)

* Bug fix from @anerokhi

* Fix floating point conversion

* Bug fix from @anerokhi and @shahor02

* Better comments

* Avoid trivial assignment

* Better comments

* Corrected behavior for first bunches

* Clang format

anerokhi committed:
Make building without tests possible again (#6699)

matthiasrichter committed:
Adding benchmark for DPLRawPageOrganizer

With on average 4 sequences of variable length per page the binary and
forward search give the following result. The parameter denotes the
number of raw pages in one message. 16 messages are emulated with
random sequence length.
--------------------------------------------------------------------------
Benchmark                                   Time           CPU Iterations
--------------------------------------------------------------------------
BM_DPLRawPageSequencerBinary/64          2726 ns       2725 ns     316405
BM_DPLRawPageSequencerBinary/512         5313 ns       5313 ns     142040
BM_DPLRawPageSequencerBinary/1024        6564 ns       6564 ns     100000
BM_DPLRawPageSequencerForward/64        21699 ns      21699 ns      32390
BM_DPLRawPageSequencerForward/512      339023 ns     339022 ns       1945
BM_DPLRawPageSequencerForward/1024    1036609 ns    1036533 ns        580

mconcas committed:
Alice 3: embedding FT3 inner disks in BP vacuum

nburmaso committed:
Rework filling process for MC track labels

Commit:
FIT: update from FEE, new field in raw data

ktf committed:
DPL: use old value for maxMemory (#6702)

ktf committed:
DPL: add driver service for configuration

ktf committed:
DPL: add driverStart callback to ServiceSpec

ktf committed:
DPL: use C++20 aggregate initialisers for services (#6701)

Simplifies too much to still be negleted, at least in backend code.

sevdokim committed:
Fix RawWriter & proper CCDB handling in Digitizer

njacazio committed:
PWGPP: apply reco. ev. cut on tracks as well (#6706)

akalweit committed:
Preparing nuclei-task for pilot testbeam and MC validation (#6698)

mfasDa committed:
[EMCAL-701] Add protection for missing / empty timeframes

Implementation according to AliceO2Group/AliceO2#5786

sawenzel committed:
Less verbose output from analysis-validation task

This was leading to extremely large log-files. If some screen information is
needed, please consider making the output outside of the loop or only
when debugging.

shahor02 committed:
CTFs written *.root.part file and renamed to *.root on file close

in order to avoid premature migration to EOS

sawenzel committed:
O2HitMerger: Fix memory leaks

* Fix a few memory leaks by using unique_ptr;
* also change a few vector->at to operator[] for
  faster access

mconcas committed:
Launch 10x more kernels in Async benchmark

As they are way more faster than sync execution we launch 10 times more to strive for more stable results

ktf committed:
DPL: websocket RFC requires encoding using network byte order (#6710)

preghenella committed:
Allow to set continueMode for GeneratorFromO2Kine

fgrosa committed:
Fix pT tolerance application in preselection

bazinski committed:
partial fix for charges

clang

shahor02 committed:
ignore values <=1 in log-normal fit

shahor02 committed:
Refurbished TPC-ITS afterburner

njacazio committed:
AO2D: extend writing of tracks and particles (#6691)

* Fix the way TOF is written

* Add PDG codes of heavier particles

* Add shift in momentum energy hack

* Add more fixes

* Add dummy TOF EventTime maker

* Resolve extra division

* Fix for sqrt(2) in FPE protection hack

mfasDa committed:
[EMCAL-703] Add optional settable subspecification for raw to digits output

- Add option to define subspecification for the output
  specs of the raw to cell converter (default: 0) and
  propagate it into the class in order to send to the
  proper subspecification in the snapshot function
- Steer subspecification via the workflow option
  subspecification

Needed in order to run the workflow on the two FLPs
in parallel. In case of running offline or on the EPN the
default subspecification (0) should be used.

matthiasrichter committed:
DPL RootTreeWriter: always amend branch name(s) if the callback is configured

So far, branch names have only been amended for branch definitions handling more
than one branch. The more logical behavior is to amend whenever a callback is
configured.

bazinski committed:
make tracklethcheader optional in various forms

preghenella committed:
Fix fatal bit in detector field

ktf committed:
DPL: avoid warning about uninitialised usage

Still not clear to me wether the condition can actually be triggered.

It might explain phantom channels to the input proxy.

jgrosseo committed:
Self index columns (#6688)

Self indexing columns development by @aalkin
Changes of isPrimaryParticle everywhere

Commit:
[o2femtodream] Analysis task for track-track femto (#6686)

- Introducing
  o PairCleaner - clean up clones
  o CutCulator - resolves the bit-wise encoding of syst. vars.
  o further improvements and bug fixes
  o plenty of documentation

- so long, and thanks for all the fish!

hjbossi committed:
[EMCAL-565]: Modified BoostHistogramUtils after testing. (#6462)

Modified BoostHistogramUtils after testing.

mconcas committed:
Add more granularity to options

ktf committed:
DPL: populate DataTakingContext from data (#6731)

Commit:
Corrupted raw data handling improved; Impact on PHOS implemented; Alignment from surway2019 updated

anerokhi committed:
Remove unintentional virtual specifier

anerokhi committed:
Use ClassDefNV when there are no virtual functions

anerokhi committed:
GPUTRDTracker: Use CAMath::Abs

anerokhi committed:
PWGPP/qaEfficiency: Prevent expensive copying

davidrohr committed:
GPU: Fix compiler warning

davidrohr committed:
FST: In the conversion tool, write to stdout even if InfoLogger is available

knopers8 committed:
[QC-626] Data Sampling: Support again varying subspecs in different parts

davidrohr committed:
DPL: Forward more FMQ command line option

davidrohr committed:
FST: Need longer delay after creating DD DATA SHM segment

davidrohr committed:
FST: Fix NUMA balancing issue, need to mlock memory directly after segment creation once

anerokhi committed:
Ignore clangd .cache folder (#6741)

davidrohr committed:
GPU Workflow: make sure to create lock file with correct permissions also when process umask is set

victor-gonzalez committed:
Next step towards the simulation challenge (#6734)

* Next step towards the simulation challenge

Support for systems without centrality/multiplicity classes
Support for pp systems and the base for pPb, Pbp and XeXe
Support for MC reco analysis

* Comparison with configuration strings correct implementation

sevdokim committed:
Added check for valid geometry addresses

ktf committed:
DPL Analysis: sanitize writing of tables

* Fix memory leaks
* Fix deletion of non-owned memory
* Fix memory corruption due to the fact wrong costructor is used
  for the an array of boolean.

davidrohr committed:
GPU: Build benchmark only for O2 builds

davidrohr committed:
GPU Display: don't draw clusters multiple times if cluster 2 collision association not available

shahor02 committed:
add ITS layers patten to TrkClusRef from AfterBurner

shahor02 committed:
Account ITS-TPC AfterBurner tracks in AOD producer

jgrosseo committed:
Use slice and array (#6629)

ktf committed:
DPL: automatically load objecs from CCDB (#6755)

mario6829 committed:
Implementation of the ITS Half Barrels (#6584)

* Implementation of the ITS Half Barrels

* Fixing ITSMisaligner macro for new method names with Half Barrels

* Do not set Layer as alignable volume

* Fixed number of half barrels

* Revert "Fixed number of half barrels"

This reverts commit 4f52866c8f191f23ed22dea4feb2717927adbb29.

* Fixed number of half barrels

* Correctly add half barrel number to hit generation and chip index number

Co-authored-by: Mario Sitta <Mario.Sitta@cern.ch>

ktf committed:
DPL: do not add a dummy sink if TFN is used

ktf committed:
Fix formatting of headers

ktf committed:
DPL: make sure single reader, multiple spawners works

jgrosseo committed:
Shorthand isPrimaryParticle (#6749)

* Shorthand isPrimaryParticle
* MC example with accessing particles and tracks
* Remove MC.h from library

aalkin committed:
DPL Analysis: Introducing conditional expressions (#6753)

`ifnode(condition, then, else)` operation is added to expressions. These can be nested and all three arguments can be arbitrary valid expressions. `condition` needs to have boolean result, `then` and `else` should return similar types (ideally the same - both floats, or both boolean, etc.). A `conditionalExpressions.cxx` tutorial example is added (note that it uses bitwise operations in filter expression and thus will only work as is with arrow > 3).

pbuehler committed:
Added script which allows automatic update of O2 data model documentation (#6762)

shahor02 committed:
median/MAD estimator + more robust log-normal fit

akalweit committed:
Several small bugfixes (mostly histogram binning and filling) (#6748)

mpuccio committed:
Task based central trigger processor (#6758)

* New task for the central event filter processor

* Task configurables from table descriptions

ddobrigk committed:
Add dynamic columns for daughter prongs (#6690)

sartozza committed:
Fixing the ProducerTask for V0, starting to implement the filling QA for daughters
